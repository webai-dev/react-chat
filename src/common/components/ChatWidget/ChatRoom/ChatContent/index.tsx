import React, { useRef, useState, useEffect, useContext } from "react";
import { makeStyles } from "@material-ui/styles";
import ChatContentItem from "./ChatContentItem";
import Box from "@material-ui/core/Box";
import ChatService from "../../../../services/chat";
import { ActiveUser, ChatRoom, RoomDetail } from "../../../../../types/types";
import { CompanyContext } from "../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../hooks/useChatRooms";
import { RoomListsContext } from "../../../../hooks/useRoomLists";
import { getRoomDetail } from "../../../../services/rooms";
import { MeContext } from "../../../../hooks/useMe";

import One2OneReadSign from "./One2OneReadSign";
import GroupReadSign from "./GroupReadSign";

// Global styles
export const useStyles = makeStyles({
  root: {
    overflowY: "auto",
    justifyContent: "flex-end",
    height: "320px",
    padding: "5px 15px",
    bottom: "75px",
  },
  loading: {
    fontSize: "12px",
    color: "#555555",
    letterSpacing: "0px",
    textAlign: "center",
  },
  loadingWrapper: {
    position: "absolute",
    top: "0",
    left: "0",
    width: "100%",
    height: "100%",
    backgroundColor: "white",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  reactionsLeft: {
    display: "flex",
    flexDirection: "column",
    alignSelf: "flex-start",
  },
  reactionsRight: {
    display: "flex",
    flexDirection: "column",
    alignSelf: "flex-end",
  },
  chatItemWrapper: {
    paddingTop: 15,
  },
  readUsers: {
    marginTop: "5px",
    marginRight: "5px",
    width: "100%",
    height: "14px",
    display: "flex",
    justifyContent: "flex-end",
    fontSize: 13,
    cursor: 'pointer',
  },
  readUsersImage: {
    borderRadius: "50%",
    border: "1px solid #cccccc",
  },
});

interface ChatContentProps {
  chatRoom: ChatRoom | null | undefined;
  connectionStatus: number;
}

/**
 * ChatContent component
 * @param {ChatContentProps} props Properties
 * @description ChatContent component consists of list of ChatContentItem components
 */
const ChatContent: React.FC<ChatContentProps> = ({
  chatRoom = null,
  connectionStatus = 1,
}) => {
  // Default styling
  const classes = useStyles();
  const { markAsRead } = ChatService;
  const { companies, setCompanies } =
    useContext(CompanyContext);
  const { loadPrevPage, loadNextPage, initializeChatRoom, updateChatRooms } =
    useContext(ChatRoomsContext);
  const { me } = useContext(MeContext);
  const {
    activeRoomList,
    companyRoomLists,
    getRoomItemById,
    updateCompanyRoomLists,
  } = useContext(RoomListsContext);
  const messageRoomRef = useRef() as React.MutableRefObject<HTMLInputElement>;

  const [detailLoading, setDetailLoading] = useState<boolean>(false);

  useEffect(() => {
    if (messageRoomRef && messageRoomRef.current && chatRoom && chatRoom.newMessageFromMe) {
      chatRoom.newMessageFromMe = false;
      messageRoomRef.current.scrollTop = messageRoomRef.current.scrollHeight;
      updateChatRooms(chatRoom);
    }
  }, [chatRoom?.newMessageFromMe, messageRoomRef]);

  useEffect(() => {
    if (messageRoomRef && messageRoomRef.current && chatRoom && chatRoom.newMessageFromOther) {
      chatRoom.newMessageFromOther = false;
      if (chatRoom.visible && chatRoom.focused && chatRoom.expanded && messageRoomRef.current.scrollTop + messageRoomRef.current.clientHeight + 73 > messageRoomRef.current.scrollHeight - 10) {
        messageRoomRef.current.scrollTop = messageRoomRef.current.scrollHeight;
      }
      updateChatRooms(chatRoom);
    }
  }, [chatRoom?.newMessageFromOther, messageRoomRef]);

  useEffect(() => {
    if (chatRoom && messageRoomRef.current) {
      if (messageRoomRef.current.scrollTop + messageRoomRef.current.clientHeight > messageRoomRef.current.scrollHeight - 10) {
        chatRoom.scrollAtBottom = true;
      } else {
        chatRoom.scrollAtBottom = false;
      }
      updateChatRooms(chatRoom);

    }
  }, [chatRoom, messageRoomRef.current?.scrollTop]);

  useEffect(() => {
    if (me && chatRoom && !chatRoom.initialized) {
      initializeChatRoom(chatRoom.roomId).then(() => {
        setDetailLoading(true);
        getRoomDetail(chatRoom.companyId, chatRoom.roomId).then(
          (data: RoomDetail | null) => {
            chatRoom.participants = data?.participants;
            chatRoom.roomType = data?.roomType;

            const participant = chatRoom.participants?.find(p => p.id === me?.id);
            const lastMessage = chatRoom.messages[chatRoom.messages.length - 1];

            if(participant && lastMessage && participant.lastReadMessageID === lastMessage.id) {
              messageRoomRef.current.scrollTop = messageRoomRef.current.scrollHeight;
              chatRoom.scrollPosition = messageRoomRef.current.scrollHeight;
            } else if (participant) {
              const message = document.getElementById(
                  `message-${participant.lastReadMessageID}`
              );

              if (message && messageRoomRef.current) {
                messageRoomRef.current.scrollTop = message.offsetTop;
                chatRoom.scrollPosition = message.offsetTop;
              }
            }

            updateChatRooms(chatRoom);

            if (messageRoomRef.current) {
              const { scrollHeight } = messageRoomRef.current;
              if (scrollHeight <= 330 && chatRoom) {
                markRoomRead(chatRoom);
              }
            }
            setDetailLoading(false);
          }
        );
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chatRoom, me]);

  useEffect(() => {
    if (chatRoom && chatRoom.initialized) {
      messageRoomRef.current.scrollTop = chatRoom.scrollPosition || 0;
    } // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chatRoom]);

  const markRoomRead = async (chatRoom: ChatRoom) => {
    const company = companies.find((c) => c.id === chatRoom.companyId);
    const room = companyRoomLists
      .map((roomList) => roomList.rooms)
      .flat()
      .find((room) => room && room.id === chatRoom.roomId);

    if (chatRoom.focused && chatRoom.expanded && chatRoom.visible) {
      if (activeRoomList && activeRoomList.rooms) {
        const room = activeRoomList.rooms.find(
          (room) => room && room.id === chatRoom.roomId
        );

        if (room && chatRoom.messages && chatRoom.messages.length > 0) {
          room.lastReadMessageID = chatRoom.messages[
            chatRoom.messages.length - 1
          ]
            ? chatRoom.messages[chatRoom.messages.length - 1].id
            : -1;
          room.unreadCount = 0;
        }

        updateCompanyRoomLists({ ...activeRoomList });
      }
      if (company && room) {
        let totalUnreadCount = 0;
        companyRoomLists.forEach((room) => {
          if (room.companyId === company.id) {
            room.rooms &&
              room.rooms.forEach((r) => {
                if (r.unreadCount > 0) {
                  totalUnreadCount += 1;
                }
              });
          }
        });

        company.unreadCount = totalUnreadCount;
        localStorage.setItem("companies", JSON.stringify([...companies]));
        setCompanies([...companies]);
      }
      await markAsRead(chatRoom.companyId, chatRoom.roomId);
    }
  };

  const handleScroll = async () => {
    if (chatRoom && messageRoomRef.current) {
      const { scrollTop, clientHeight, scrollHeight } = messageRoomRef.current;

      chatRoom.scrollPosition = scrollTop;
      if (chatRoom.prevHasMore && !chatRoom.loadingPrev && scrollTop === 0) {
        loadPrevPage(chatRoom.roomId).then((messages) => {
          if (messageRoomRef.current) {
            messageRoomRef.current.scrollTop = messages.length * 63 + 1;
          }
        });
      } else if (
        chatRoom.nextHasMore &&
        !chatRoom.loadingNext &&
        scrollTop + clientHeight === scrollHeight
      ) {
        const room = getRoomItemById(chatRoom.roomId);
        if (
          chatRoom.messages[chatRoom.messages.length - 1] &&
          room?.lastReadMessageID !==
            chatRoom.messages[chatRoom.messages.length - 1].id
        ) {
          const company = companies.find(
            (c) => c && c.id === chatRoom.companyId
          );
          const room = companyRoomLists
            .map((roomList) => roomList.rooms)
            .flat()
            .find((room) => room && room.id === chatRoom.roomId);
          if (chatRoom.focused && chatRoom.expanded && chatRoom.visible) {
            if (activeRoomList && activeRoomList.rooms) {
              const room = activeRoomList.rooms.find(
                (room) => room && room.id === chatRoom.roomId
              );

              if (room) {
                room.lastReadMessageID = chatRoom.messages[
                  chatRoom.messages.length - 1
                ]
                  ? chatRoom.messages[chatRoom.messages.length - 1].id
                  : -1;
                room.unreadCount = 0;
              }

              updateCompanyRoomLists({ ...activeRoomList });
            }
            if (company && room) {
              let totalUnreadCount = 0;
              companyRoomLists.forEach((room) => {
                if (room.companyId === company.id) {
                  room.rooms &&
                    room.rooms.forEach((r) => {
                      if (r && r.unreadCount > 0) {
                        totalUnreadCount += 1;
                      }
                    });
                }
              });

              company.unreadCount = totalUnreadCount;

              localStorage.setItem("companies", JSON.stringify([...companies]));
              setCompanies([...companies]);
            }

            await markAsRead(chatRoom.companyId, chatRoom.roomId);
          }
        }

        loadNextPage(chatRoom.roomId).then((messages) => {
          if (messageRoomRef.current) {
            messageRoomRef.current.scrollTop =
              messageRoomRef.current.scrollHeight - messages.length * 63 - 1;
          }
        });
        chatRoom.messages.forEach((message) => (message.new = false));
      } else if (scrollTop + clientHeight === scrollHeight) {
        markRoomRead(chatRoom);
        // await markAsRead(chatRoom.companyId, chatRoom.roomId);
      }
    }
  };

  const participant = chatRoom?.participants?.find((participant: ActiveUser) => participant.id !== me?.id);
  const singleReadPosition = participant && chatRoom?.roomType !== 0 ? participant.lastReadMessageID : -1;
  let groupReadPosition: any = {};
  let groupAllReadPosition = -1;

  if (chatRoom && chatRoom.roomType === 0) {
    chatRoom.participants?.forEach((participant: ActiveUser) => {
      const messageID = participant.lastReadMessageID ? participant.lastReadMessageID.toString() : '-1';

      if (messageID !== '-1' && participant.id !== me?.id) {
        if (groupReadPosition[messageID]) {
          groupReadPosition[messageID].push(participant.id);
        } else {
          groupReadPosition[messageID] = [participant.id];
        }
      }
    });

    const sortedIds = Object.keys(groupReadPosition).map(id => parseInt(id)).sort((a, b) => a > b ? 1 : -1);
    let sum = 0;

    for (let i = 0; i < sortedIds.length; i++) {
      sum += groupReadPosition[sortedIds[i]].length;
      if (sum === ((chatRoom.participants?.length || 1) - 2)) {
        groupAllReadPosition = sortedIds[i];
        break;
      }
    }
  }

  return (
    <>
      <div
        data-testid="chat-content"
        className={classes.root}
        ref={messageRoomRef}
        onScroll={handleScroll}
      >
        {(chatRoom?.loadingPrev && detailLoading) && (
          <Box
            display="flex"
            flex={1}
            height={90}
            justifyContent="center"
            alignItems="center"
          >
            <p className={classes.loading}>Loading history...</p>
          </Box>
        )}
        {chatRoom &&
          (chatRoom.messages || [])
            .map((message, index) => (
              <React.Fragment>
                <div
                  id={`message-${message.id || message.trackingID}`}
                  key={index}
                  className={classes.chatItemWrapper}
                >
                  {
                     detailLoading && (
                        <Box className={classes.loadingWrapper}>
                          <p className={classes.loading}>Loading history...</p>
                        </Box>
                     )
                  }
                  <ChatContentItem
                    message={message}
                    connectionStatus={connectionStatus}
                    roomId={chatRoom.roomId}
                    companyId={chatRoom.companyId ? chatRoom.companyId : null}
                    singleReadPosition={singleReadPosition}
                    groupReadPosition={groupReadPosition}
                    groupAllReadPosition={groupAllReadPosition}
                    roomType={chatRoom.roomType}
                  />
                </div>
                {
                  (chatRoom?.roomType !== 0 && message.id === singleReadPosition) && (
                    <One2OneReadSign
                      chatRoom={chatRoom}
                      me={me}
                    />
                  )
                }
                {
                  (chatRoom?.roomType === 0 && groupReadPosition[message.id]) && (
                    <GroupReadSign
                      participants={chatRoom?.participants}
                      groupReadPosition={groupReadPosition}
                      message={message}
                    />
                  )
                }
              </React.Fragment>
            ))}
        {chatRoom?.loadingNext && (
          <Box
            display="flex"
            flex={1}
            height={90}
            justifyContent="center"
            alignItems="center"
          >
            <p className={classes.loading}>Loading history...</p>
          </Box>
        )}
      </div>
    </>
  );
};

export default ChatContent;
