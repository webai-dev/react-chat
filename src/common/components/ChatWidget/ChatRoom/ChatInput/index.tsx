import React, {useState, useContext, useRef, useEffect} from "react";
import { Container } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import SendIcon from "@material-ui/icons/Send";
import chat from "../../../../services/chat";
import { ChatRoom } from "../../../../../types/types";
import { MeContext } from "../../../../hooks/useMe";
import { CompanyContext } from "../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../hooks/useChatRooms";
import { RoomListsContext } from "../../../../hooks/useRoomLists";
import { v4 as uuidv4 } from "uuid";
import InputEmoji from "react-input-emoji";
import { NewMessage } from "../../../../../types";

// Global styles
export const useStyles = makeStyles({
  icon: {
    width: "31px",
    height: "31px",
    padding: "0 10px",
    cursor: "pointer",
  },
  input: {
    outline: "none",
    height: "39px",
    width: "100%",
    padding: "0 10px",
    border: "1px solid #D6D6D6",
  },
  root: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    width: "100%",
    height: "55px",
    padding: "5px 15px",

    color: "#3870A0",
    bottom: 0,
    fontSize: "24px",
  },
});

interface ChatInputProps {
  chatRoom: ChatRoom | null | undefined;
  connectionStatus: number;
  setSendFailed: (flag: boolean) => void;
}

/**
 * ChatInput component
 * @param {ChatInputProps} props Properties
 * @description ChatInput component consists of input box and send button
 */
const ChatInput: React.FC<ChatInputProps> = ({
  chatRoom = null,
  connectionStatus = 1,
  setSendFailed,
}) => {
  // Default styling
  const classes = useStyles();
  const [message, setMessage] = useState<string>("");
  const [isSending, setIsSending] = useState<boolean>(false);
  const { sendMessage } = chat;
  const { me } = useContext(MeContext);
  const { selectedCompany } = useContext(CompanyContext);
  const { updateChatRooms } = useContext(ChatRoomsContext);
  const { companyRoomLists } = useContext(RoomListsContext);
  const inputRef = useRef() as React.MutableRefObject<HTMLInputElement>;

  const handleChange = (e: string) => {
    setMessage(e);
  };

  // useEffect(() => {
  //   console.log(connectionStatus);
  // }, [connectionStatus]);

  const handleSend = async () => {
    let data = {
      message: message,
      trackingID: uuidv4(),
      sentAt: new Date(),
    };
    setIsSending(true);
    setMessage("");
    if (chatRoom && me) {
      chatRoom.messages = [
        ...chatRoom.messages,
        {
          id: 0,
          message: message,
          messageType: 0,
          readBy: me.id,
          sentAt: data.sentAt.toString(),
          trackingID: data.trackingID,
          user: {
            id: me.id,
            firstName: me.firstName,
            lastName: me.lastName,
            isOnline: true,
            companyMembership: [],
          },
          new: true,
          check: "sent",
        },
      ];
      chatRoom.newMessageFromMe = true;

      updateChatRooms(chatRoom);
    }
    let updatedCompanyRooms = companyRoomLists.find(
      (company) => selectedCompany && company.companyId === selectedCompany.id
    );
    if (updatedCompanyRooms && updatedCompanyRooms.rooms) {
      let targetRoom = updatedCompanyRooms.rooms.find(
        (r) => chatRoom && r && r.id === chatRoom.roomId
      );
      if (targetRoom) {
        updatedCompanyRooms.rooms = [
          targetRoom,
          ...updatedCompanyRooms?.rooms.filter(
            (r) => chatRoom && r && r.id !== chatRoom.roomId
          ),
        ];
      }
    }
    if (connectionStatus === 1) {
      const response = await sendMessage(
        selectedCompany && selectedCompany.id,
        chatRoom?.roomId,
        data
      );
      if (!response && chatRoom?.messages) {
        chatRoom.messages = chatRoom?.messages.map((m) =>
          m.id === 0 ? { ...m, check: "failed" } : m
        );
      }
    } else {
      let response: NewMessage | null = null;
      if (!(connectionStatus === 1)) {
        response = await sendMessage(
            selectedCompany && selectedCompany.id,
            chatRoom?.roomId,
            data
        );

        if (response) {
          if (chatRoom?.messages) {
            chatRoom.messages = chatRoom?.messages.map((m) =>
                m.id === 0 ? { ...m, check: "received" } : m
            );
            updateChatRooms(chatRoom);
          }
        } else {
          let started = Date.now();
          // loop every 10 seconds
          let interval = setInterval(async function () {
            if (Date.now() - started > 46000) {
              clearInterval(interval);
              if (chatRoom?.messages) {
                chatRoom.messages = chatRoom?.messages.map((m) =>
                    m.id === 0 ? { ...m, check: "failed" } : m
                );
                setSendFailed(true);
              }
            }
            // Retry sending message for first 2 minutes or after 5 minutes
            else {
              // let response = connectionStatus === 1 ? 1 : 0;
              if (response) {
                if (chatRoom?.messages) {
                  chatRoom.messages = chatRoom?.messages.map((m) =>
                      m.id === 0 ? { ...m, check: "received" } : m
                  );
                  updateChatRooms(chatRoom);
                }
                setSendFailed(false);
                clearInterval(interval);
              } else {
                response = await sendMessage(
                    selectedCompany && selectedCompany.id,
                    chatRoom?.roomId,
                    data
                );
              }
            }
          }, 15000);
        }

      }
    }
    setIsSending(false);
  };

  useEffect(() => {
    if(isSending)
      inputRef.current.focus();
  }, [inputRef.current]);

  return (
    <>
      <Container data-testid="chat-input" className={classes.root}>
        <InputEmoji
          value={message}
          className={classes.input}
          onChange={handleChange}
          cleanOnEnter
          onEnter={() => {
            handleSend();
            // inputRef.current.autofocus = true;
          }}
          placeholder="Message"
          ref={inputRef}
        />

        <SendIcon onClick={handleSend} className={classes.icon} />
      </Container>
    </>
  );
};

export default ChatInput;
