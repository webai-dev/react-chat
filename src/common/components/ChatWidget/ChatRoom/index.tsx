import React, { useRef, useEffect, useContext, useState } from "react";
import { makeStyles } from "@material-ui/styles";
import Accordion from "@material-ui/core/Accordion";
import AccordionSummary from "@material-ui/core/AccordionSummary";
import AccordionDetails from "@material-ui/core/AccordionDetails";
import ChatRoomHeader from "./ChatRoomHeader";
import ChatInput from "./ChatInput";
import ChatContent from "./ChatContent";
import { CompanyContext } from "../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../hooks/useChatRooms";
import { RoomListsContext } from "../../../hooks/useRoomLists";

// Global styles
export const useStyles = makeStyles({
  root: {
    display: "flex",
    flexDirection: "column",
    width: "370px",
    height: "462px",
    position: "absolute",
    boxShadow: "0px 3px 6px #0000000A",
    background: "#fff",
    bottom: 0,
    right: (index) => ((index as number) + 1) * 390,
    margin: 0,
    padding: 0,
  },
  accordion: {
    display: "flex",
    flexDirection: "column",
    width: "370px",
    position: "fixed",
    boxShadow: "0px 3px 6px #0000000A",
    background: "#fff",
    bottom: 0,
    right: (index) => ((index as number) + 1) * 390,
    margin: 0,
    padding: 0,
    borderRadius: "0 !important",
  },
  accordionDetails: {
    display: "block",
    padding: 0,
  },
  accordionExpanded: {
    margin: "0 !important",
  },
  accordionSummary: {
    paddingLeft: 16,
    paddingRight: 0,
  },
  accordionSummaryUnread: {
    paddingLeft: 16,
    paddingRight: 0,
    backgroundColor: "#3870A0",
  },
  accordionSummaryContent: {
    margin: "0 !important",
    display: "block",
  },
  expandIconMinimized: {
    color: "#ffffff",
    paddingRight: "10px",
    transform: "none",
  },
  expandIcon: {
    color: "#a9b4bb",
    transform: "none",
    paddingRight: "10px",
  },
  messageContainer: {
    position: "relative",
    height: "100%",
  },
  hr: {
    height: "0.1px",
    margin: "0 auto",
    width: "90%",
    background: "#707070",
    opacity: 0.5,
    alignSelf: "center",
    marginBottom: "5px",
  },
});

interface ChatRoomProps {
  roomId: number;
  index: number;
  connectionStatus: number;
}

/**
 * ChatRoom component
 * @param {ChatRoomProps} props Properties
 * @description ChatRoom component consists of sections - ChatRoomHeader, ChatContent and ChatInput
 */
const ChatRoom: React.FC<ChatRoomProps> = ({
  roomId = -1,
  index = 0,
  connectionStatus = 1,
}) => {
  const { chatRooms, updateChatRooms, expandChatRoom } =
    useContext(ChatRoomsContext);
  const { companyRoomLists } = useContext(RoomListsContext);

  const chatRoom =
    chatRooms && chatRooms.find((room) => room.roomId === roomId);
  const { selectedCompany } = useContext(CompanyContext);
  const roomItem =
    companyRoomLists &&
    companyRoomLists
      .map((roomList) => roomList.rooms)
      .flat()
      .find((r) => r && r.id === chatRoom?.roomId);

  const [sendFailed, setSendFailed] = useState(false);

  // Default styling
  const classes = useStyles(index);

  // States
  const wrapperRef = useRef<HTMLInputElement | null>(null);
  useEffect(() => {
    /**
     * Alert if clicked on outside of chat room
     */
    function handleClickOutside(event: any) {
      if (wrapperRef.current && !wrapperRef.current.contains(event.target)) {
        if (chatRoom) {
          chatRoom.focused = false;
          updateChatRooms(chatRoom);
        }
      }
    }
    // Bind the event listener
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [wrapperRef, chatRoom, updateChatRooms]);

  useEffect(() => {
    const onLoseFocus = () => {
      if (chatRoom && "hidden" in document) {
        chatRoom.focused = false;
        updateChatRooms(chatRoom);
      }
    };

    document.addEventListener("visibilitychange", onLoseFocus);

    return () => {
      document.removeEventListener("visibilitychange", onLoseFocus);
    };
  }, [wrapperRef, chatRoom, updateChatRooms]);


  return (
    <>
      <Accordion
        data-testid="chat-room"
        ref={wrapperRef}
        expanded={chatRoom && chatRoom.expanded}
        classes={{
          root: classes.accordion,
          expanded: classes.accordionExpanded,
        }}
        onClick={() => {
          if (chatRoom) {
            chatRoom.focused = true;
            updateChatRooms(chatRoom);
          }
        }}
        onChange={(event: React.ChangeEvent<{}>, expand: boolean) => {
          if (chatRoom) {
            chatRoom.expanded = expand;
            expandChatRoom(
              chatRoom.roomId,
              selectedCompany ? selectedCompany.id : null,
              expand
            );
            updateChatRooms(chatRoom);
          }
        }}
      >
        <AccordionSummary
          classes={{
            root:
              chatRoom &&
              (!chatRoom.expanded || !chatRoom.focused || !chatRoom.scrollAtBottom) &&
              roomItem &&
              roomItem.unreadCount > 0
                ? classes.accordionSummaryUnread
                : classes.accordionSummary,
            content: classes.accordionSummaryContent,
          }}
        >
          <ChatRoomHeader
            chatRoom={chatRoom}
            connectionStatus={connectionStatus}
          />
        </AccordionSummary>
        <hr className={classes.hr} />
        <AccordionDetails className={classes.accordionDetails}>
          <div className={classes.messageContainer}>
            <ChatContent
              chatRoom={chatRoom}
              connectionStatus={connectionStatus}
            />
            <ChatInput
              chatRoom={chatRoom}
              setSendFailed={(flag) => setSendFailed(flag)}
              connectionStatus={connectionStatus}
            />
          </div>
        </AccordionDetails>
      </Accordion>
    </>
  );
};

export default ChatRoom;
