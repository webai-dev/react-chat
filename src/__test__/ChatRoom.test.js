import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ChatRoom from "../common/components/ChatWidget/ChatRoom";
import renderer from "react-test-renderer";

// react-emoji package error in test
// it("renders without crashing", () => {
//   const { getByTestId } = render(<ChatRoom />);
//   const chatWidget = getByTestId("chat-room");
//   expect(chatWidget).toBeTruthy();
// });

it("matches snapshot", () => {
  let index = 1,
    roomId = 1,
    setShowModal = () => undefined;
  const tree = renderer
    .create(
      <ChatRoom index={index} roomId={roomId} setShowModal={setShowModal} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
