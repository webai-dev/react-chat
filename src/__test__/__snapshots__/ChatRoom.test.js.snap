// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiPaper-root MuiAccordion-root makeStyles-accordion-2 makeStyles-accordion-13 MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  data-testid="chat-room"
  onClick={[Function]}
>
  <div
    aria-disabled={false}
    aria-expanded={false}
    className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-accordionSummary-5"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiAccordionSummary-content makeStyles-accordionSummaryContent-7"
    >
      <div
        data-testid="chat-room-header"
      >
        <div
          className="MuiContainer-root makeStyles-root-14 MuiContainer-maxWidthLg"
        >
          <div
            className="makeStyles-detailsContainer-16"
          >
            <div
              className="makeStyles-imageContainer-18"
            />
            <div
              className="makeStyles-info-19"
            >
              <div
                className="makeStyles-name-20 makeStyles-name-25"
              >
                
              </div>
              <div
                className="makeStyles-department-21 makeStyles-department-26"
              >
                
              </div>
            </div>
          </div>
          <div
            className="makeStyles-iconContainer-17 makeStyles-iconContainer-24"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              onClick={[Function]}
              viewBox="0 0 24 24"
            >
              <path
                d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm-9-2V7H4v3H1v2h3v3h2v-3h3v-2H6zm9 4c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root makeStyles-closeIcon-15"
              focusable="false"
              onClick={[Function]}
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiCollapse-container MuiCollapse-hidden"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      >
        <div
          role="region"
        >
          <hr
            className="makeStyles-hr-11"
          />
          <div
            className="MuiAccordionDetails-root makeStyles-accordionDetails-3"
          >
            <div
              className="makeStyles-messageContainer-10"
            >
              <div
                className="makeStyles-root-27"
                data-testid="chat-content"
                onScroll={[Function]}
              >
                <br />
              </div>
              <div
                className="MuiContainer-root makeStyles-root-36 MuiContainer-maxWidthLg"
                data-testid="chat-input"
              >
                <div
                  className="react-emoji"
                >
                  <div
                    className="react-emoji-picker--container"
                  >
                    <div
                      className="react-emoji-picker--wrapper"
                    >
                      <div
                        className="react-emoji-picker"
                      />
                    </div>
                  </div>
                  <div
                    className="react-input-emoji--container"
                    style={
                      Object {
                        "borderColor": "#EAEAEA",
                        "borderRadius": 21,
                        "fontFamily": "sans-serif",
                        "fontSize": 15,
                      }
                    }
                  >
                    <div
                      className="react-input-emoji--wrapper"
                      onClick={[Function]}
                    >
                      <div
                        className="react-input-emoji--placeholder"
                      >
                        Message
                      </div>
                      <div
                        className="react-input-emoji--input"
                        contentEditable={true}
                        onBlur={[Function]}
                        onKeyDown={[Function]}
                        tabIndex={0}
                      />
                    </div>
                  </div>
                  <button
                    className="react-input-emoji--button"
                    onClick={[Function]}
                  >
                    <svg
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12 0C5.373 0 0 5.373 0 12s5.373 12 12 12 12-5.373 12-12S18.627 0 12 0m0 22C6.486 22 2 17.514 2 12S6.486 2 12 2s10 4.486 10 10-4.486 10-10 10"
                      />
                      <path
                        d="M8 7a2 2 0 1 0-.001 3.999A2 2 0 0 0 8 7M16 7a2 2 0 1 0-.001 3.999A2 2 0 0 0 16 7M15.232 15c-.693 1.195-1.87 2-3.349 2-1.477 0-2.655-.805-3.347-2H15m3-2H6a6 6 0 1 0 12 0"
                      />
                    </svg>
                  </button>
                </div>
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root makeStyles-icon-34"
                  focusable="false"
                  onClick={[Function]}
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
